File: Ebay/3_Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        map = {}
        res = 0
        
        for i in range(len(s)):
            if s[i] not in map: # Means the current character has not been existed on the current substring 
                map[s[i]] = 1   # Add the unique character into substring     
            else: # Duplicated character
                while map[s[i]] != 0:
                    map[s[left]] -= 1
                    left += 1
                map[s[i]] = 1
            res = max(res, i-left+1)
        return res

File: PastYears/Practice_Algorithm/3. Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = s.length();
        int l = 0, res = 0;
        Map<Character, Integer>m = new HashMap<>();
        for(int i = 0; i < len; i++){
            Character key = s.charAt(i);
            if(m.containsKey(key)){
                Integer pos = m.get(key);
                if(pos >= l){
                    res = Math.max(res, i-l);  
                    l = pos+1;
                }else{
                   res = Math.max(res, i-l+1); 
                }
                
            }else{
                res = Math.max(res, i-l+1);
            }
            m.put(key,i);
        }
        return res;
    }
}
